File: backend/src/respira_ally/core/schemas/daily_log.py
"""
Daily Log Schemas
Pydantic models for Daily Log API endpoints
"""
from datetime import date, datetime, timedelta
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


# ============================================================================
# Base Schemas
# ============================================================================

class DailyLogBase(BaseModel):
    """Base daily log information (all fields optional except log_date)"""
    log_date: date = Field(..., description="Date of the log entry (ONLY required field)")
    medication_taken: bool | None = Field(None, description="Whether medication was taken (NULL = not recorded)")
    water_intake_ml: int | None = Field(None, ge=0, le=10000, description="Water intake in milliliters (NULL = not recorded)")
    exercise_minutes: int | None = Field(None, ge=0, le=480, description="Exercise duration in minutes (0-480, max 8 hours)")
    smoking_count: int | None = Field(None, ge=0, le=100, description="Number of cigarettes smoked (COPD risk factor)")
    symptoms: str | None = Field(None, max_length=500, description="Reported symptoms")
    mood: Literal["GOOD", "NEUTRAL", "BAD"] | None = Field(None, description="Patient mood")

    @field_validator("log_date")
    @classmethod
    def validate_log_date(cls, v: date) -> date:
        """
        Validate log date is not in the future and within reasonable past range

        Rules:
        - Cannot be future date
        - Cannot be older than 365 days
        """
        today = date.today()

        if v > today:
            raise ValueError("Log date cannot be in the future")

        oldest_allowed = today - timedelta(days=365)
        if v < oldest_allowed:
            raise ValueError("Log date cannot be older than 365 days")

        return v

    @field_validator("water_intake_ml")
    @classmethod
    def validate_water_intake(cls, v: int | None) -> int | None:
        """
        Validate water intake for COPD patients

        Warning ranges:
        - <500ml: Dehydration risk (COPD patients need adequate hydration)
        - >5000ml: Excessive intake (normal: 2000-3000ml)
        """
        if v is None:
            return v

        if v < 500:
            # Note: This is a warning, not an error. We allow the value but log it.
            # In production, this could trigger a notification to therapist
            pass
        elif v > 5000:
            # Similarly, high intake is allowed but should be monitored
            pass

        return v

    @field_validator("exercise_minutes")
    @classmethod
    def validate_exercise_minutes(cls, v: int | None) -> int | None:
        """
        Validate exercise duration for COPD patients

        Warning ranges:
        - <10 minutes: Insufficient activity (COPD patients need regular exercise)
        - >120 minutes: Excessive activity (may strain COPD patients)

        Recommended: 30-60 minutes of moderate exercise daily
        """
        if v is None:
            return v

        if v < 10:
            # Low activity warning
            pass
        elif v > 120:
            # High activity warning (potential over-exertion)
            pass

        return v

    @field_validator("smoking_count")
    @classmethod
    def validate_smoking_count(cls, v: int | None) -> int | None:
        """
        Validate smoking count (CRITICAL for COPD management)

        ANY smoking is a risk factor for COPD progression!

        Alert ranges:
        - ≥1 cigarette: Should trigger therapist notification for cessation support
        - ≥10 cigarettes: High-risk behavior, immediate intervention needed
        """
        if v is None:
            return v

        if v >= 1:
            # CRITICAL: Any smoking should alert therapist
            # In production, this should trigger immediate notification
            pass

        return v

    @field_validator("symptoms")
    @classmethod
    def validate_symptoms(cls, v: str | None) -> str | None:
        """
        Validate symptoms and check for critical keywords

        Critical keywords that should alert therapist:
        - "無法呼吸", "不能呼吸", "呼吸困難"
        - "胸痛", "胸悶"
        - "暈倒", "昏倒", "失去意識"
        - "咳血", "吐血"
        """
        if v is None:
            return v

        critical_keywords = [
            "無法呼吸", "不能呼吸", "呼吸困難", "喘不過氣",
            "胸痛", "胸悶", "心臟痛",
            "暈倒", "昏倒", "失去意識", "昏迷",
            "咳血", "吐血", "血痰"
        ]

        v_lower = v.lower()
        for keyword in critical_keywords:
            if keyword.lower() in v_lower:
                # In production, this should trigger immediate alert to therapist
                # For now, we just validate the data is stored
                pass

        return v


# ============================================================================
# Request Schemas
# ============================================================================

class DailyLogCreate(DailyLogBase):
    """
    Daily Log Creation Request
    Patients submit daily health logs via LIFF
    """
    patient_id: UUID = Field(..., description="Patient user ID")


class DailyLogUpdate(BaseModel):
    """
    Daily Log Update Request
    All fields optional for partial updates
    """
    medication_taken: bool | None = None
    water_intake_ml: int | None = Field(None, ge=0, le=10000)
    exercise_minutes: int | None = Field(None, ge=0, le=480)
    smoking_count: int | None = Field(None, ge=0, le=100)
    symptoms: str | None = Field(None, max_length=500)
    mood: Literal["GOOD", "NEUTRAL", "BAD"] | None = None


# ============================================================================
# Response Schemas
# ============================================================================

class DailyLogResponse(DailyLogBase):
    """
    Daily Log API Response
    Returns complete daily log information
    """
    log_id: UUID = Field(..., description="Unique log ID")
    patient_id: UUID = Field(..., description="Patient user ID")
    created_at: datetime = Field(..., description="Log creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    model_config = ConfigDict(from_attributes=True)


class DailyLogListResponse(BaseModel):
    """
    Daily Log List Response (with pagination)
    """
    items: list[DailyLogResponse]
    total: int = Field(..., description="Total number of logs")
    page: int = Field(..., description="Current page (0-indexed)")
    page_size: int = Field(..., description="Items per page")
    has_next: bool = Field(..., description="Has next page")


# ============================================================================
# Query Schemas
# ============================================================================

class DailyLogQuery(BaseModel):
    """
    Query parameters for filtering daily logs
    """
    patient_id: UUID | None = Field(None, description="Filter by patient ID")
    start_date: date | None = Field(None, description="Filter logs from this date (inclusive)")
    end_date: date | None = Field(None, description="Filter logs until this date (inclusive)")
    mood: Literal["GOOD", "NEUTRAL", "BAD"] | None = Field(None, description="Filter by mood")
    medication_taken: bool | None = Field(None, description="Filter by medication compliance")


# ============================================================================
# Statistics Schemas
# ============================================================================

class DailyLogStats(BaseModel):
    """
    Daily log statistics for a patient
    """
    total_logs: int = Field(..., description="Total number of logs")
    medication_adherence_rate: float = Field(..., ge=0, le=100, description="Medication adherence percentage")
    avg_water_intake_ml: float | None = Field(None, description="Average daily water intake (NULL if no data)")
    avg_exercise_minutes: float | None = Field(None, description="Average daily exercise duration in minutes")
    avg_smoking_count: float | None = Field(None, description="Average daily cigarette count (COPD risk tracking)")
    mood_distribution: dict[str, int] = Field(..., description="Count of each mood type")
    date_range: dict[str, date] = Field(..., description="Start and end dates of logs")

File: frontend/liff/src/api/daily-log.ts
/**
 * Daily Log API - LIFF Daily Health Log Submission
 * Mock 模式支援前端獨立開發
 */

import { apiClient, isMockMode } from '../services/api-client'
import {
  type DailyLogCreate,
  type DailyLogResponse,
  type DailyLogListResponse,
  Mood,
} from '../types/daily-log'

// ============================================================================
// Mock Data
// ============================================================================

const MOCK_LOG_RESPONSE: DailyLogResponse = {
  log_id: 'mock-log-id-123',
  patient_id: 'mock-patient-id-456',
  log_date: new Date().toISOString().split('T')[0],
  medication_taken: true,
  water_intake_ml: 2000,
  exercise_minutes: 30,
  smoking_count: 0,
  symptoms: '輕微咳嗽，無其他不適',
  mood: Mood.GOOD,
  created_at: new Date().toISOString(),
  updated_at: new Date().toISOString(),
}

const MOCK_LOGS: DailyLogResponse[] = [
  {
    log_id: 'log-1',
    patient_id: 'mock-patient-id-456',
    log_date: '2025-10-20',
    medication_taken: true,
    water_intake_ml: 2000,
    exercise_minutes: 30,
    smoking_count: 0,
    symptoms: '輕微咳嗽',
    mood: Mood.GOOD,
    created_at: '2025-10-20T08:30:00Z',
    updated_at: '2025-10-20T08:30:00Z',
  },
  {
    log_id: 'log-2',
    patient_id: 'mock-patient-id-456',
    log_date: '2025-10-19',
    medication_taken: true,
    water_intake_ml: 1800,
    exercise_minutes: 20,
    smoking_count: 2,
    symptoms: '呼吸順暢',
    mood: Mood.GOOD,
    created_at: '2025-10-19T09:00:00Z',
    updated_at: '2025-10-19T09:00:00Z',
  },
  {
    log_id: 'log-3',
    patient_id: 'mock-patient-id-456',
    log_date: '2025-10-18',
    medication_taken: false,
    water_intake_ml: 1500,
    exercise_minutes: 0,
    smoking_count: 5,
    symptoms: '呼吸急促，輕微胸悶',
    mood: Mood.NEUTRAL,
    created_at: '2025-10-18T10:15:00Z',
    updated_at: '2025-10-18T10:15:00Z',
  },
]

// ============================================================================
// API Functions
// ============================================================================

export const dailyLogApi = {
  /**
   * 提交每日日誌（LIFF 表單）
   */
  async createLog(data: DailyLogCreate): Promise<DailyLogResponse> {
    if (isMockMode) {
      // 模擬延遲 (600-1200ms)
      await new Promise((resolve) =>
        setTimeout(resolve, 600 + Math.random() * 600)
      )

      console.log('[MOCK] POST /daily-logs', data)

      // 驗證必填欄位
      if (!data.log_date || data.medication_taken === undefined || !data.water_intake_ml) {
        throw new Error('必填欄位不可為空')
      }

      // 驗證範圍
      if (data.water_intake_ml < 0 || data.water_intake_ml > 10000) {
        throw new Error('飲水量必須在 0-10000 毫升之間')
      }

      if (data.exercise_minutes !== undefined && data.exercise_minutes !== null) {
        if (data.exercise_minutes < 0 || data.exercise_minutes > 480) {
          throw new Error('運動分鐘數必須在 0-480 之間')
        }
      }

      if (data.smoking_count !== undefined && data.smoking_count !== null) {
        if (data.smoking_count < 0 || data.smoking_count > 100) {
          throw new Error('吸菸支數必須在 0-100 之間')
        }
      }

      // 返回 Mock 回應
      return {
        ...MOCK_LOG_RESPONSE,
        ...data,
        log_id: `log-${Date.now()}`,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      }
    }

    // 真實 API 調用
    const response = await apiClient.post<DailyLogResponse>('/daily-logs', data)
    return response
  },

  /**
   * 取得病患的日誌列表（分頁）
   */
  async getLogs(params: {
    patient_id: string
    page?: number
    page_size?: number
  }): Promise<DailyLogListResponse> {
    if (isMockMode) {
      await new Promise((resolve) =>
        setTimeout(resolve, 600 + Math.random() * 600)
      )

      console.log('[MOCK] GET /daily-logs', params)

      const page = params.page || 0
      const pageSize = params.page_size || 20

      return {
        items: MOCK_LOGS,
        total: MOCK_LOGS.length,
        page,
        page_size: pageSize,
        has_next: false,
      }
    }

    // 真實 API 調用
    const response = await apiClient.get<DailyLogListResponse>('/daily-logs', {
      params,
    })
    return response
  },

  /**
   * 取得特定日期的日誌
   */
  async getLogByDate(
    patient_id: string,
    log_date: string
  ): Promise<DailyLogResponse | null> {
    if (isMockMode) {
      await new Promise((resolve) =>
        setTimeout(resolve, 300 + Math.random() * 300)
      )

      console.log('[MOCK] GET /daily-logs/by-date', { patient_id, log_date })

      const log = MOCK_LOGS.find((l) => l.log_date === log_date)
      return log || null
    }

    // 真實 API 調用
    const response = await apiClient.get<DailyLogResponse | null>(
      `/daily-logs/by-date`,
      {
        params: { patient_id, log_date },
      }
    )
    return response
  },
}
